File: app/src/main/java/kztproject/jp/splacounter/repository/AuthRepository.java
package kztproject.jp.splacounter.repository;

import javax.inject.Inject;

import io.reactivex.Completable;
import kztproject.jp.splacounter.api.TodoistClient;
import kztproject.jp.splacounter.model.User;
import kztproject.jp.splacounter.model.UserResponse;
import kztproject.jp.splacounter.preference.AppPrefs;
import kztproject.jp.splacounter.preference.AppPrefsProvider;

public class AuthRepository {

    private final TodoistClient client;
    private final AppPrefsProvider prefs;

    @Inject
    public AuthRepository(TodoistClient client, AppPrefsProvider prefs) {
        this.client = client;
        this.prefs = prefs;
    }

    public Completable login(String inputString) {
        return client.getUser(inputString)
                .flatMapCompletable(this::save);
    }

    private Completable save(UserResponse response) {
        User user = response.getUser();
        AppPrefs schema = prefs.get();
        schema.putUserId(user.getId());
        schema.putUserName(user.getFullName());
        return Completable.complete();
    }
}

File: app/src/main/java/kztproject/jp/splacounter/repository/AuthRepository.kt
package kztproject.jp.splacounter.repository

import io.reactivex.Completable
import kztproject.jp.splacounter.api.RewardListClient
import kztproject.jp.splacounter.api.TodoistClient
import kztproject.jp.splacounter.model.RewardUser
import kztproject.jp.splacounter.model.UserResponse
import kztproject.jp.splacounter.preference.PrefsWrapper
import javax.inject.Inject

class AuthRepository @Inject constructor(private val todoistClient: TodoistClient,
                                         private val rewardListClient: RewardListClient) {

    fun login(inputString: String): Completable {
        return todoistClient.getUser(inputString)
                .flatMap { response: UserResponse -> rewardListClient.findUser(response.user.id) }
                .flatMapCompletable { save(it) }
    }

    fun signUp(todoistToken: String): Completable {
        return todoistClient.getUser(todoistToken)
                .flatMap { response -> rewardListClient.createUser(response.user.id.toLong()) }
                .flatMapCompletable { save(it) }
    }

    private fun save(user: RewardUser): Completable {
        PrefsWrapper.userId = user.id
        return Completable.complete()
    }
}
